### PR TITLE
sql/schemachanger: Enable declarative create index with hash sharded indexes

### PR DESCRIPTION
**Pull request by fqazi**
1. Add split and scatter support for declarative CREATE INDEX and refactor code and adopt in the legacy schema changer
2. Enable creating hash shared indexes in the declarative schema changer

Release notes: None
Epic: none

### CODE REVIEW

**Review by rharding6373 — APPROVED** [review]
I reviewed the sql files and :lgtm: with just some nits.

> Reviewed 6 of 72 files at r1, 1 of 53 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-:-NL2B8ry32f-4TPR_iSH:b-lftlt6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`-- commits` line 18 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NL2AcZKDOYxkLbdQwaO):*
nit: s/shraded/sharded
___
*[`-- commits` line 21 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NL2AhOM3XRF_cAjzlRl):*
nit: s/heck/check
___
*[`pkg/sql/index_split_scatter.go` line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NL21kXyDDJZ37R2QE12:-NL21kXz0i0ThknPQKFs:b1cjet5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbc1bb03cdccf4bfe969d34ca4062866793229fc/pkg/sql/index_split_scatter.go#L35)):*
> ```Go
> }
> 
> // NewIndexSplitAndScatter creates a new scexec.IndexSpanSplitter implemetation.
> ```

nit: s/implemetation/implementation
___
*[`pkg/sql/index_split_scatter.go` line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NL299NX9gWEGFZWUBTC:-NL299NX9gWEGFZWUBTD:bb88stv) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbc1bb03cdccf4bfe969d34ca4062866793229fc/pkg/sql/index_split_scatter.go#L48)):*
> ```Go
> 	ctx context.Context, table catalog.TableDescriptor, indexToBackfill catalog.Index,
> ) error {
> 	if is.shouldSplitAndScatter(indexToBackfill) {
> ```

nit: This call to `shouldSplitAndScatter` is redundant since it's called immediately in `MaybeSplitIndexSpansForPartitioning`, too. 
___
*[`pkg/sql/index_split_scatter.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NL22F869GdtyAZzlOBB:-NL22F869GdtyAZzlOBC:b-f1bg04) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbc1bb03cdccf4bfe969d34ca4062866793229fc/pkg/sql/index_split_scatter.go#L55)):*
> ```Go
> 	}
> 	// Only perform splits on the system tenant.
> 	if !is.codec.ForSystemTenant() || indexToBackfill.IsTemporaryIndexForBackfill() {
> ```

nit: consider moving this check first in this function, to avoid returning an unnecessary error and doing unnecessary work
___
*[`pkg/sql/index_split_scatter.go` line 84 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NL24XaT2EH0CJV557d0:-NL24XaT2EH0CJV557d1:b-hoa2oi) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbc1bb03cdccf4bfe969d34ca4062866793229fc/pkg/sql/index_split_scatter.go#L84)):*
> ```Go
> 	}
> 	const backfillSplitExpiration = time.Hour
> 	expirationTime := is.db.Clock().Now().Add(backfillSplitExpiration.Nanoseconds(), 0) // Iterate through all partitioning lists to get all possible list
> ```

formating nit: move comment to new line

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go [hunk 1]
```diff
@@ -658,7 +683,32 @@ func addColumnsForSecondaryIndex(
...
+				}
+			}
+			checkConstraintBucketValues.WriteString(")")
+			_, err = parser.ParseExpr(checkConstraintBucketValues.String())
+			if err != nil {
+				panic(err)
+			}
+			shardCheckConstraint := &scpb.CheckConstraint{
+				TableID:      idxSpec.secondary.TableID,
+				ConstraintID: idxSpec.secondary.ConstraintID + 2,
```
Comment (postamar) [inline pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:703]: Can we avoid relying on magical offsets here? Also why isn't this using `b.NextTableConstraintID`?

File: pkg/sql/logictest/testdata/logic_test/crdb_internal_catalog [hunk 2]
```diff
@@ -159,7 +159,7 @@ SELECT id, strip_volatile(descriptor) FROM crdb_internal.kv_catalog_descriptor
...
..., "type": {"family": "IntFamily", "oid": 20, "width": 64}}], "dependsOn": [111], "formatVersion": 3, "id": 112, "indexes": [{"constraintId": 2, "createdExplicitly": true, "foreignKey": {}, "geoConfig": {}, "id": 2, "interleave": {}, "keyColumnDirections": ["ASC"], "keyColumnIds": [2], "keyColumnNames": ["v"], "keySuffixColumnIds": [3], "name": "idx", "partitioning": {}, "sharded": {}, "version": 4}], "isMaterializedView": true, "name": "mv", "nextColumnId": 4, "nextConstraintId": 4, "nextIndexId": 4, "nextMutationId": 1, "parentId": 106, "primaryIndex": {"constraintId": 1, "encodingType": 1, "foreignKey": {}, "geoConfig": {}, "id": 1, "interleave": {}, "keyColumnDirections": ["ASC"], "keyColumnIds": [3], "keyColumnNames": ["rowid"], "name": "mv_pkey", "partitioning": {}, "sharded": {}, "storeColumnIds": [1, 2], "storeColumnNames": ["k", "v"], "unique": true, "version": 4}, "privileges": {"ownerProto": "root", "users": [{"privileges": 2, "userProto": "admin", "withGrantOption": 2}, {"privileges": 2, "userProto": "root", "withGrantOption": 2}], "version": 2}, "replacementOf": {"time": {}}, "unexposedParentSchemaId": 107, "version": "10", "viewQuery": "SELECT k, v FROM db.public.kv"}}
```
Comment (postamar) [inline pkg/sql/logictest/testdata/logic_test/crdb_internal_catalog:162]: Something's wrong here, the largest constraint ID in this descriptor is 2 yet NextConstraintID is 4.

**Review by postamar — CHANGES_REQUESTED** [review]
Here's my partial review. I'm very suspicious of at least some of the constraint ID changes in the test outputs.

**Review by fqazi — COMMENTED** [review]
Yes, that part was buggy it should be fixed now.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-:-NLb9N2246SUtgHDvU4l:b-yeyz1u)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @postamar and @rharding6373)

___
*[`pkg/sql/index_split_scatter.go` line 55 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NL22F869GdtyAZzlOBB:-NLbSrqyDQ0FiOtdaBrM:b-6wbqxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbc1bb03cdccf4bfe969d34ca4062866793229fc/pkg/sql/index_split_scatter.go#L55)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

nit: consider moving this check first in this function, to avoid returning an unnecessary error and doing unnecessary work
</blockquote></details>

We still need to split along the partitioning in that case, so that part is fine.
___
*[`pkg/sql/index_split_scatter.go` line 84 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NL24XaT2EH0CJV557d0:-NLb980ZEkxSDWj2jwAU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbc1bb03cdccf4bfe969d34ca4062866793229fc/pkg/sql/index_split_scatter.go#L84)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

formating nit: move comment to new line
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 699 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NKxgZvZ76w2YUWytD-5-r4-699:-NLb996tCSkI8UyJLqpo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75b5bcbded0c8e6cab480865ada7d77bfdae9701/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L699)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This error should be handled as an assertion failure.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 703 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NKxgZvZ76w2YUWytD-5-r4-703:-NLb99bT0XwTh6GGDXD8:b1burkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/75b5bcbded0c8e6cab480865ada7d77bfdae9701/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L703)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Can we avoid relying on magical offsets here? Also why isn't this using `b.NextTableConstraintID`?
</blockquote></details>

Yes, good point.

Done.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go` line 711 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NKxgZvZ76w2YUWytD-5-r4-711:-NLb9HAW7Yz-beQ-WCNW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/75b5bcbded0c8e6cab480865ada7d77bfdae9701/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go#L711)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Would it make sense to move all of this into `maybeCreateAndAddShardCol`?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/crdb_internal_catalog` line 162 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/94703#-NKzHluwinuP70ZiGVdw-r4-162:-NLb9IN48YZkEexgMcYF:b9eeejs) ([raw file](https://github.com/cockroachdb/cockroach/blob/75b5bcbded0c8e6cab480865ada7d77bfdae9701/pkg/sql/logictest/testdata/logic_test/crdb_internal_catalog#L162)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Something's wrong here, the largest constraint ID in this descriptor is 2 yet NextConstraintID is 4.
</blockquote></details>

Ah figured out that we are incorrectly incrementing the constraint ID extra times. This is now sane.

**Comment by postamar** [comment]
I'll go through this with fresh eyes once it's rebased on https://github.com/cockroachdb/cockroach/pull/95631 which is a very noisy change.

**Comment by fqazi** [comment]
Let me take rebase this on top

**Comment by fqazi** [comment]
@postamar This is ready for another look

File: pkg/sql/schemachanger/scop/immediate_mutation_visitor_generated.go [hunk 3]
```diff
@@ -473,3 +474,8 @@ func (op AddColumnToIndex) Visit(ctx context.Context, v ImmediateMutationVisitor
 func (op RemoveColumnFromIndex) Visit(ctx context.Context, v ImmediateMutationVisitor) error {
 	return v.RemoveColumnFromIndex(ctx, op)
 }
+
+// Visit is part of the MutationOp interface.
+func (op MaybeAddSplitForIndex) Visit(ctx context.Context, v MutationVisitor) error {
+	return v.MaybeAddSplitForIndex(ctx, op)
+}
```
Comment (postamar) [inline pkg/sql/schemachanger/scop/immediate_mutation_visitor_generated.go:481]: Seems like you need to `dev gen go` again in the first commit.

File: pkg/sql/schemachanger/scplan/internal/opgen/opgen_secondary_index.go [hunk 4]
```diff
@@ -28,6 +28,12 @@ func init() {
 						IsSecondaryIndex: true,
 					}
 				}),
+				emit(func(this *scpb.SecondaryIndex) *scop.MaybeAddSplitForIndex {
+					return &scop.MaybeAddSplitForIndex{
+						TableID: this.TableID,
+						IndexID: this.IndexID,
+					}
+				}),
```
Comment (postamar) [inline pkg/sql/schemachanger/scplan/internal/opgen/opgen_secondary_index.go:36]: Should we be emitting this op when creating new primary indexes also?

File: pkg/sql/schemachanger/scexec/scmutationexec/dependencies.go [hunk 5]
```diff
@@ -115,4 +116,10 @@ type DeferredMutationStateUpdater interface {
 
 	// RefreshStats refresh stats for a given descriptor.
 	RefreshStats(id descpb.ID)
+
+	// AddIndexForMaybeSplitAndScatter adds an index to maybe split and scatter,
+	// if it makes sense.
```
Comment (postamar) [inline pkg/sql/schemachanger/scexec/scmutationexec/dependencies.go:121]: reword comment

File: pkg/sql/schemachanger/scexec/dependencies.go [hunk 6]
```diff
@@ -222,6 +222,9 @@ type IndexSpanSplitter interface {
 
 	// MaybeSplitIndexSpans will attempt to split the backfilled index span.
 	MaybeSplitIndexSpans(ctx context.Context, table catalog.TableDescriptor, indexToBackfill catalog.Index) error
+
+	// MaybeSplitIndexSpansForPartitioning will attempt to split the backfilled index span.
```
Comment (postamar) [inline pkg/sql/schemachanger/scexec/dependencies.go:226]: perhaps explain how it differs from MaybeSplitIndexSpans?

File: pkg/sql/schemachanger/scdeps/sctestdeps/backfill.go [hunk 7]
```diff
@@ -127,3 +127,9 @@ func (s *indexSpanSplitter) MaybeSplitIndexSpans(
 ) error {
 	return nil
 }
+
+func (s *indexSpanSplitter) MaybeSplitIndexSpansForPartitioning(
+	_ context.Context, _ catalog.TableDescriptor, _ catalog.Index,
+) error {
+	return nil
+}
```
Comment (postamar) [inline pkg/sql/schemachanger/scdeps/sctestdeps/backfill.go:135]: move this while IndexSpanSplitter implementation to another source file than `backfill.go`.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go [hunk 8]
```diff
@@ -182,6 +183,11 @@ func CreateIndex(b BuildCtx, n *tree.CreateIndex) {
 	// Picks up any geoconfig parameters, hash sharded one are
 	// picked independently.
 	maybeApplyStorageParameters(b, n, &idxSpec)
+	// Assign the secondary constraint ID now, since we may have added a check
+	// constraint earlier.
+	if idxSpec.secondary.IsUnique {
+		idxSpec.secondary.ConstraintID = b.NextTableConstraintID(idxSpec.secondary.TableID)
+	}
```
Comment (postamar) [inline pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:190]: any interest in behaving the same way for non-unique indexes also?

**Review by postamar — APPROVED** [review]
LGTM modulo a few things which should be straightforward.

File: pkg/sql/schemachanger/scplan/internal/opgen/opgen_secondary_index.go (hunk 4 above)
Comment (fqazi) [reply pkg/sql/schemachanger/scplan/internal/opgen/opgen_secondary_index.go:36]: Good point this looks like a regression, let me add this in the primary index code path too.

File: pkg/sql/schemachanger/scexec/scmutationexec/dependencies.go (hunk 5 above)
Comment (fqazi) [reply pkg/sql/schemachanger/scexec/scmutationexec/dependencies.go:121]: Reworded to: 	// AddIndexForMaybeSplitAndScatter splits and scatters rows for a given index,
	                // if it's either hash sharded or under the system tenant.

File: pkg/sql/schemachanger/scexec/dependencies.go (hunk 6 above)
Comment (fqazi) [reply pkg/sql/schemachanger/scexec/dependencies.go:226]: // MaybeSplitIndexSpans will attempt to split the backfilled index span, if
	// the index is in the system tenant or is partitioned.

	// MaybeSplitIndexSpansForPartitioning will split backfilled index spans
	// across hash-sharded index boundaries if applicable.

File: pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go (hunk 8 above)
Comment (fqazi) [reply pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go:190]: They shouldn't have a constraint ID assigned. So, I think this is correct behaviorally:

	// Used within the table descriptor to uniquely identify individual
	// constraints, which is only set for primary keys and unique secondary
	// indexes.

**Comment by fqazi** [comment]
@postamar TFTR!

File: pkg/sql/schemachanger/scdeps/sctestdeps/backfill.go (hunk 7 above)
Comment (fqazi) [reply pkg/sql/schemachanger/scdeps/sctestdeps/backfill.go]: Done.

**Pull request merged**